Added ConditionalTask

Dr15Jones opened the pull request:
#### PR description:

A ConditionalTask will automatically add modules in the ConditionalTask to the associated Path in the proper order based on consumes/produce calls of the modules in the Path and in the ConditionalTask. If no module on the Path directly or indirectly uses a module in the ConditionalTask, that module is set to unscheduled execution.

#### PR validation:

All framework unit tests pass and newly added unit tests also pass.

Dr15Jones committed:
Improve testing modules

- added fillDescriptions
- fixed a bug when filter not supposed to produce a data product
- allow testing for missing data products

Dr15Jones committed:
Added ConditionalTask object

Dr15Jones committed:
Added code to handle ConditionalTasks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37305/28942

- This PR adds an extra 132KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- DataFormats/Provenance (**core**)
- FWCore/Framework (**core**)
- FWCore/ParameterSet (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@makortel, @wddgit, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -21,6 +22,7 @@
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 #include "FWCore/ParameterSet/interface/Registry.h"
 #include "FWCore/ServiceRegistry/interface/PathContext.h"
+#include "FWCore/Reflection/interface/DictionaryTools.h"
```
probably not needed

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -36,6 +38,7 @@
 #include <list>
 #include <map>
 #include <exception>
+#include <unordered_set>
```
Should be in .h

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -378,6 +381,154 @@ namespace edm {
+          (ci.process().empty() or ci.process() == processConfiguration->processName())) {
+        auto productModuleLabel = ci.label();
+        if (productModuleLabel.empty()) {
+          for (auto const& branch : conditionalModuleBranches) {
```
probably should add a comment saying this is a consumesMany

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -389,6 +540,61 @@ namespace edm {
+    auto itCondBegin = std::find(modnames.begin(), modnames.end(), "#");
+
+    std::unordered_set<std::string> conditionalmods;
+    //need to capture
```
Either complete this thought or remove.

Dr15Jones commented:
I plan to address my own comments once the intitial tests finish.

makortel commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -2848,6 +2947,193 @@ def testTask(self):
+            testTask1.visit(visitor)
+            self.assertEqual(l, set(['mesproducer', 'mproducer', 'mproducer2', 'mfilter', 'd', 'messource']))
+            l2 = testTask1.moduleNames
+            self.assertEqual(l, set(['mesproducer', 'mproducer', 'mproducer2', 'mfilter', 'd', 'messource']))
```
Should this be
```suggestion
            self.assertEqual(l2, set(['mesproducer', 'mproducer', 'mproducer2', 'mfilter', 'd', 'messource']))
```
?

makortel commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -2848,6 +2947,193 @@ def testTask(self):
+            testTask1.add(testTask3)
+            process.myTask1 = testTask1
+
+            # test the validation that occurs when attaching a Task to a Process
```
```suggestion
            # test the validation that occurs when attaching a ConditionalTask to a Process
```

makortel reviewed (commented):
Went through the python part

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ec433d/23282/summary.html
**COMMIT**: 76938a5410a4ce07620828bee725b26189cbc8dd
**CMSSW**: CMSSW_12_4_X_2022-03-22-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/37305/23282/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test materialBudgetTrackerPlots had ERRORS
</pre>

## Comparison Summary



There are some workflows for which there are errors in the baseline:
 534.0 step 1
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

The workflows 1001.0, 1000.0, 136.88811, 136.874, 136.8311, 136.793, 136.7611, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /pool/condor/dir_63795/jenkins/workspace/compare-root-files-short-matrix/data/PR-ec433d/39434.911_TTbar_14TeV+2026D88_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3695650
 - DQMHistoTests: Total failures: 52023
 - DQMHistoTests: Total nulls: 62
 - DQMHistoTests: Total successes: 3643543
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 204 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-22-1100+ec433d/49074/triggerResults): no differences found

Martin-Grunewald commented:
@Dr15Jones thanks for this.
I assume we need a backport to 12_3 as this is the release used for initial data taking.
@silviodonato

makortel reviewed (commented):
Went through the C++ code

makortel commented on FWCore/ParameterSet/python/SequenceTypes.py:
```diff
@@ -1626,51 +1670,136 @@ def remove(self, something):
 
+class Task(_TaskBase) :
+    """Holds EDProducers, EDFilters, ESProducers, ESSources, Services, and Tasks.
+    A Task can be associated with Sequences, Paths, EndPaths and the Schedule.
```
Right,  but it can be associated with `ConditionalTask`. Or would some other wording than "associated" be more precise for the relationship between `Task` and `ConditionalTask`?

Dr15Jones committed:
Improvements based on review comments

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -378,6 +381,154 @@ namespace edm {
+                                                 TypeID(itBranch->second->wrappedType().typeInfo()),
+                                                 itBranch->second->className())) {
+                          productFromConditionalModule = true;
+                          break;
```
I think you are correct.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37305/28971

- This PR adds an extra 132KB to repository

cmsbuild commented:
Pull request #37305 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -378,6 +381,154 @@ namespace edm {
+            }
+
+            auto condWorker =
+                getWorker(productModuleLabel, proc_pset, workerManager_, preg, prealloc, processConfiguration);
```
Good catch. There was no test for consumesMany.

Dr15Jones replied:
Test added

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -378,6 +381,154 @@ namespace edm {
+                                                           prealloc,
+                                                           processConfiguration);
+            returnValue.insert(returnValue.end(), dependents.begin(), dependents.end());
+            returnValue.push_back(condWorker);
```
You missed the fact that as tryToPlaceConditionalModules it slowly removes conditional modules from the list. Therefore what you'd actually get is
```
a b f1 c f2
```

makortel replied:
Ah right, the insertion is being done here only if `productFromConditionalModule == true`. For non-alias branches the logic is clear. But for the alias branches, I think, the alias-to module could be inserted many times. (or did I miss something again?)

In the consumesMany branch I see elements being removed from `conditionalModules`, but there is no other action done with it. I believe that would lead to duplicate entries in the Path.

Dr15Jones replied:
Fixed for consumesMany

Dr15Jones commented:
Setting `process.options.wantSummary = True` gives

```
TimeReport ---------- Modules in Path: p ---[Real sec]----
TimeReport  per event    per visit  Name
TimeReport   0.012894     0.012894  f1
TimeReport   0.013273     0.013273  prodOnPath
TimeReport   0.001023     0.001023  f2
TimeReport   0.000091     0.000091  f3
TimeReport   0.000783     0.000783  #
TimeReport   0.000589     0.000589  a
TimeReport   0.001828     0.001828  b
TimeReport   0.000098     0.000098  c
TimeReport   0.000000     0.000000  d
TimeReport   0.000000     0.000000  e
TimeReport   0.000000     0.000000  f
TimeReport   0.000000     0.000000  @
TimeReport  per event    per visit  Name
```

Which is not correct

Dr15Jones commented:
So the TriggerNamesService poses a problem. It assumes the name order in the ParameterSet corresponding to the name of a path IS the order of the modules. That is clearly not the case for ConditionalTasks.

Dr15Jones commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @Dr15Jones
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

Dr15Jones commented:
I think we are caught in a dilemma. We do not want ConditionalTask to cause the top level PSet hash to change in the case a module dependency changes between two jobs running the same configuration (that is how the SwitchProducer behaves). We also need to record the module labels associated to each trigger bit. Right now (I think) the assumption is the PSet associated to the path name is the order (which is what the TriggerNamesService assumes). 

For the presently running processes, it would be possible to correct the module order for the TriggerNamesService. The question then becomes, how would one be able to interpret the module order bits in TriggerResults stored in a file?

Dr15Jones committed:
Added ConditionalTask test using consumesMany

Dr15Jones committed:
Fix mayConsumes handling for ConditionalTasks

Use proper module label and avoid reusing a module.

Dr15Jones committed:
Improved handling of regular consumes for ConditionalTasks

- avoid unnecessary check on alias finding
- break out of loop properly

Dr15Jones committed:
A module's path bit position can be different from processing order

The bit position of a module onto a path does not have to be the same as the modules order to be run on the path.
This is needed to allow ConditionalTasks to be added to a path in a set run order but still be able to use the unaltered parameter value holding the list of module labels in order to later find the module labels.
All ConditionalTask's modules are added to the end of the parameter and are bracketed by two non-module labels [# and @]. Modules added explicitly to the path are at the front of the parameter and in their run order.
The TriggerNamesService knows how to translate the path bit position into the proper module label stored in the parameter.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37305/29043

- This PR adds an extra 160KB to repository

Dr15Jones commented:
please test

cmsbuild commented:
Pull request #37305 was updated. @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
unhold

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ec433d/23469/summary.html
**COMMIT**: 290304617c55c02efa08699346d4a6568da0b8d0
**CMSSW**: CMSSW_12_4_X_2022-03-28-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37305/23469/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3589460
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3589436
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-28-1100+ec433d/49250/triggerResults): no differences found

Martin-Grunewald commented:
@Dr15Jones 

Yes, it expects i to be the ith trigger path in the menu, and j to be the jth module on that trigger path.

This could be after all transformations from the python representation to how the framework actually runs the HLTmenu. It is assumed unchanged from event to event, whether or not the jth module is actually executed or not for an event (say because an earlier filter already bailed out with false).

Sorry, what change do you have in mind?

Dr15Jones commented:
@Martin-Grunewald 
> Is there a way to get the enumeration in the order of 

I was unable to figure out a way. The problem is the following say we have this configuration
```
p = cms.Path( f, ConditionalTask(c, g))
```
For jobs running on a node with only CPU, f depends on c and not g. For jobs running on a node with a GPU, f depends on g and not c. (That is basically how the SwitchProducers work.)

A constraint we have is all jobs using the same configuration MUST generate the same ParameterSet hash (since is it used to identify related Run data). That means the CPU and GPU case above must store and use the same top level ParameterSet. Which means the PSet holding the strings related to the module labels for the Path "p" must be identical so they can not be in the run order. In addition, the only place the framework archives which modules were used by which path is in the Path's PSet. These constraints mean the framework can't use the ParameterSet system to store the actual module order it can only store the information it had before doing the dependency checking.

>However, for previous processes the TNS only provides the path list (AFAICS).

It turns out there are two functions to help with previous processes
https://github.com/cms-sw/cmssw/blob/320b54524aae2d38ae9737f0ae66bb2604da5d32/FWCore/Framework/interface/TriggerNamesService.h#L59-L72

Dr15Jones commented:
@Martin-Grunewald 

One thing to keep in mind, the EDFilters DO appear in run order and their position bits DO correspond to their placement in the Path's PSet list.

Martin-Grunewald commented:
You mention above:
``` In addition, the only place the framework archives which modules were used by which path is in the Path's PSet. ```
Is that PSet saved and available for later? Wouldn't that solve the problem?

Dr15Jones commented:
@Martin-Grunewald 

An alternative implements could be I do not store the ConditionalTask information in the regular Path PSet and instead create a new one (say "@p_conditional"). I then 'hide' the Modules which are assigned to the Path (so they do not appear in the TriggerResults and such). It does make the callbacks to Services harder to handle (since they are told expicitly if a module is on a Path or not and many Services depend on that behavior).

I though the present implementation was better since it does allow the Summary system to show the actual run order of the modules and it allows storing the case where there is an exception (which is something the Task system doesn't really handle well). This also implementation also seems to work well with all the known Services and the places which directly look at `"@trigger_paths"` is tiny (with only `HLTConfigData` presently not handling this change).

Dr15Jones commented:
@Martin-Grunewald 
> Sorry, what change do you have in mind?

I never answered this. The minimum change would be the code that dumps what modules are in a Path would have to at least skip the _process directives_ since there are no corresponding PSets stored. It could be extended to understand those _process directives_ and do something special when printing.

A more future proof change would be to use the TNS to understand the Paths and their modules (with an extension to TNS for the latter).

Martin-Grunewald commented:
Hmm OK (as long as you code the changes to TNS and HLTConfidData etc.)

Dr15Jones committed:
Update HLTConfigData to handle ConditionalTasks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37305/29068

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #37305 was updated. @Martin-Grunewald, @Dr15Jones, @smuzaffar, @makortel, @cmsbuild, @missirol can you please check and sign again.

Dr15Jones commented:
please test

makortel commented on FWCore/Framework/interface/TriggerNamesService.h:
```diff
@@ -83,6 +84,9 @@ namespace edm {
         return (modulenames_.at(i)).at(j);
       }
+      ///The label is an indicator of a path scheduling construct and not an actual module.
+      using ScheduingConstructLabelSet = std::unordered_set<std::string, std::hash<std::string>, std::equal_to<>>;
+      static ScheduingConstructLabelSet const& schedulingConstructLabels();
```
I would also **guess** that a linear search on a `vector` would be faster (even for some more labels). On the other hand this is not used in performance critical part.

Dr15Jones replied:
I’d prefer to keep it as it is as it makes its use easy.

makortel commented on HLTrigger/HLTcore/src/HLTConfigData.cc:
```diff
@@ -117,9 +118,17 @@ void HLTConfigData::extract() {
   // Obtain module labels of all modules on all trigger paths
   const unsigned int n(size());
   moduleLabels_.reserve(n);
+  std::unordered_set<std::string> processDirectives = {{"@"}, {"#"}};
```
How about asking these from `edm::service::TriggerNamesService::schedulingConstructLabels()`? `HLTrigger/HLTcore` seems to already depend on `FWCore/Framework`.

Dr15Jones replied:
I created a standalone function that all changed classes now use.

makortel reviewed (commented):
From the first look of the commits of yesterday and today. Need to take another look on the big picture.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ec433d/23513/summary.html
**COMMIT**: 8099e0a907083c9958163f4a725446f83df16a68
**CMSSW**: CMSSW_12_4_X_2022-03-29-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37305/23513/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3585896
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3585872
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-29-1100+ec433d/49281/triggerResults): no differences found

Dr15Jones committed:
Moved path configuration parsing to a standalone function

This allows sharing of this code across two cases.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37305/29090

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #37305 was updated. @Martin-Grunewald, @Dr15Jones, @smuzaffar, @makortel, @cmsbuild, @missirol can you please check and sign again.

Dr15Jones commented:
please test

makortel commented on FWCore/Utilities/interface/path_configuration.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace edm::path_configuration {
+  ///The label is an indicator of a path scheduling construct and not an actual module.
+  using ScheduingConstructLabelSet = std::unordered_set<std::string, std::hash<std::string>, std::equal_to<>>;
+  ScheduingConstructLabelSet const& schedulingConstructLabels();
```
Typo (got moved)
```suggestion
  using SchedulingConstructLabelSet = std::unordered_set<std::string, std::hash<std::string>, std::equal_to<>>;
  SchedulingConstructLabelSet const& schedulingConstructLabels();
```

Dr15Jones replied:
Done (that's what happens when one does edits while attending meetings)

makortel commented on FWCore/Utilities/src/path_configuration.cc:
```diff
@@ -0,0 +1,38 @@
+  //Takes the Parameter associated to a given Path and converts it to the list of modules
+  // in the same order as the Path's position bits
+  std::vector<std::string> configurationToModuleBitPosition(std::vector<std::string> iConfig) {
+    auto labelsToRemove = schedulingConstructLabels();
```
Maybe
```suggestion
    auto const& labelsToRemove = schedulingConstructLabels();
```
?

Dr15Jones replied:
done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37305/29092

- This PR adds an extra 284KB to repository

cmsbuild commented:
Pull request #37305 was updated. @Martin-Grunewald, @Dr15Jones, @smuzaffar, @makortel, @cmsbuild, @missirol can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ec433d/23546/summary.html
**COMMIT**: 8b992105533716be1100a6c5d94290ee645a504f
**CMSSW**: CMSSW_12_4_X_2022-03-30-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37305/23546/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3585896
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3585860
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-30-1100+ec433d/49307/triggerResults): no differences found

Dr15Jones commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -378,6 +381,154 @@ namespace edm {
+          if (itFound == conditionalModules.end()) {
+            //Check to see if this was an alias
+            auto findAlias = aliasMap.equal_range(productModuleLabel);
+            if (findAlias.first != findAlias.second) {
```
Changed

Dr15Jones commented:
Ping

missirol commented:
Thanks @makortel , @Dr15Jones .

I have started to test this PR on HLT configs, but I don't have any solid conclusions yet.

In one set of tests, I start from an HLT config which is "fully taskified" (i.e. for every path, producers are grouped in ConditionalTasks wherever possible); in that case, one hits errors that relate to the two caveats (for example, there are a few HLT paths in which some of the products are not consumed by the path itself by construction, so it looks like we can't use the current ConditionalTask there, as that would make those modules run unscheduled), and I'm still working my way through those.

I still need to check if ConditionalTasks are sufficient to achieve the minimum goal, which is to start supporting SwitchProducers with both branches (`cpu` and `cuda`) in the offline HLT configs (so, using ConditionalTasks for parts of the reconstruction that can be offloaded, not necessarily everywhere else in the HLT menus).

I wonder if you are considering backporting this in time for `12_3_0` (as HLT will use that release at the start of Run 3), or if you'd prefer to have feedback from HLT before considering backports.

makortel commented:
Thanks @missirol for a quick test!

> (for example, there are a few HLT paths in which some of the products are not consumed by the path itself by construction, so it looks like we can't use the current ConditionalTask there, as that would make those modules run unscheduled)

Could you elaborate what kind of behavior would you expect in this case? If I understood correctly, a Path would have modules that are not consumed by any other module in the Path, but they would be expected to be run as part of the Path, is that correct? 

> I still need to check if ConditionalTasks are sufficient to achieve the minimum goal, which is to start supporting SwitchProducers with both branches (`cpu` and `cuda`) in the offline HLT configs (so, using ConditionalTasks for parts of the reconstruction that can be offloaded, not necessarily everywhere else in the HLT menus).

I'd expect it to work in a case where he `cpu` and `cuda` case-modules, or the modules in ConditionalTask that are consumed only by the `cpu`/`cuda` case-modules, are not consumed outside of the Path. I _think_ anything more complex would hit to the  caveats.


> I wonder if you are considering backporting this in time for `12_3_0` (as HLT will use that release at the start of Run 3), or if you'd prefer to have feedback from HLT before considering backports.

We can backport this (and subsequent PRs) after a few IBs have confirmed that it (they) doesn't cause problems, if based on tests done in 12_4_X the backport(s) would be useful for HLT.

missirol commented:
> Could you elaborate what kind of behavior would you expect in this case? If I understood correctly, a Path would have modules that are not consumed by any other module in the Path, but they would be expected to be run as part of the Path, is that correct?

Yes. One good example is the so-called Scouting 'packers', which currently run (scheduled) in the Scouting Path, and create the products saved in the output of the scouting stream (see, for example, the sequence `HLTPixelOnlyPFScoutingPackingSequence` in the [dump of the GRun HLT menu](https://raw.githubusercontent.com/cms-sw/cmssw/48f1b480ffea6ba486a02c54281a8079c9ce7170/HLTrigger/Configuration/test/OnLine_HLT_FULL.py)). I think we also have "AlCa" paths that create output products in a similar fashion (not consumed in the Path that produces them). One possibility would keep these modules in the Path as scheduled, rather than in a ConditionalTask (Andrea B. already suggested this last week, if I understood him correctly).

> I'd expect it to work in a case where he cpu and cuda case-modules, or the modules in ConditionalTask that are consumed only by the cpu/cuda case-modules, are not consumed outside of the Path. I think anything more complex would hit to the caveats.

I made an attempt where (1) I took the GRun menu and taskified it in full (all Paths), (2) converted it into the standard `GRun_cff` file that we use for HLT tests in matrix wfs, (3) removed `AlCa_` and `Scouting` Paths via a customisation function, and (4) ran wf `11634.0`. This worked, which I think it's an improvement compared to the problem originally reported in #36938. The SwitchProducers at HLT currently only have one branch (the `cpu` one), but it seems that we are not hitting a problem with Paths that use them. Now I'm trying to figure out if the issues with the AlCa and Scouting Paths are easy to fix.

> We can backport this (and subsequent PRs) after a few IBs have confirmed that it (they) doesn't cause problems, if based on tests done in 12_4_X the backport(s) would be useful for HLT.

I'm hopeful that we are getting closer to the solution, and having this in `12_3_0` would ease further tests for us, imho.

makortel commented:
> > Could you elaborate what kind of behavior would you expect in this case? If I understood correctly, a Path would have modules that are not consumed by any other module in the Path, but they would be expected to be run as part of the Path, is that correct?
> 
> Yes. One good example is the so-called Scouting 'packers', which currently run (scheduled) in the Scouting Path, and create the products saved in the output of the scouting stream (see, for example, the sequence `HLTPixelOnlyPFScoutingPackingSequence` in the [dump of the GRun HLT menu](https://raw.githubusercontent.com/cms-sw/cmssw/48f1b480ffea6ba486a02c54281a8079c9ce7170/HLTrigger/Configuration/test/OnLine_HLT_FULL.py)). I think we also have "AlCa" paths that create output products in a similar fashion (not consumed in the Path that produces them). One possibility would keep these modules in the Path as scheduled, rather than in a ConditionalTask (Andrea B. already suggested this last week, if I understood him correctly).

Thanks. I think it would be best to keep those modules explicitly in the `Path`, because that reflects well their intended behavior. 

We can't really make the `ConditionalTask` to cause them to be run without consumption within `Path`, because then the CPU-only/CUDA-only chains of modules that are used as an input for `SwithcProducer` would be run unconditionally as well.

makortel commented:
(just to remind in case it was lost in the earlier messages, this PR needs now @cms-sw/hlt-l2 signature)

missirol commented:
> (just to remind in case it was lost in the earlier messages, this PR needs now @cms-sw/hlt-l2 signature)

I would give Martin a chance to comment on the `HLTConfigData` update, because he knows `HLTcore/` better than me. Unless there are further comments, I'll sign for HLT by end of tomorrow.

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

makortel commented:
unhold

based on https://github.com/cms-sw/cmssw/issues/36938#issuecomment-1090605271

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
@Dr15Jones there are a couple of comments from @makortel in FWCore/ParameterSet/python/SequenceTypes.py that don't look having been ever addressed or answered. Could any of you confirm that they do not require further fixes here, and that this PR can be merged as it is?

makortel commented:
@perrotta I checked again and my comments https://github.com/cms-sw/cmssw/pull/37305#discussion_r832545923 and https://github.com/cms-sw/cmssw/pull/37305#discussion_r832563357 (that are the only ones in `SequenceTypes.py` that are not "Outaded" AFAICT) have been addressed.

perrotta commented:
> @perrotta I checked again and my comments [#37305 (comment)](https://github.com/cms-sw/cmssw/pull/37305#discussion_r832545923) and [#37305 (comment)](https://github.com/cms-sw/cmssw/pull/37305#discussion_r832563357) (that are the only ones in `SequenceTypes.py` that are not "Outaded" AFAICT) have been addressed.

Thank you @makortel for the confirmation

perrotta commented:
+1
